processing/src/main/java/io/druid/query/BySegmentSkippingQueryRunner.java

```diff
@@ -38,14 +38,26 @@ public BySegmentSkippingQueryRunner(
   }
 
   @Override
-  public Sequence<T> run(Query<T> query, Map<String, Object> responseContext)
+  public Sequence<T> run(QueryPlus<T> queryPlus, Map<String, Object> responseContext)
   {
-    if (QueryContexts.isBySegment(query)) {
-      return baseRunner.run(query, responseContext);
+    if (QueryContexts.isBySegment(queryPlus.getQuery())) {
+      return baseRunner.run(queryPlus, responseContext);
     }
 
-    return doRun(baseRunner, query, responseContext);
+    return doRun(baseRunner, queryPlus, responseContext);
   }
 
-  protected abstract Sequence<T> doRun(QueryRunner<T> baseRunner, Query<T> query, Map<String, Object> context);
+  /**
+   * @deprecated override {@link #doRun(QueryRunner, QueryPlus, Map)} instead
+   */
+  @Deprecated
+  protected Sequence<T> doRun(QueryRunner<T> baseRunner, Query<T> query, Map<String, Object> context)
+  {
+    return doRun(baseRunner, QueryPlus.wrap(query), context);
+  }
+
+  protected Sequence<T> doRun(QueryRunner<T> baseRunner, QueryPlus<T> queryPlus, Map<String, Object> context)
```

Comment:
Such recursion is supposed to be broken by subclass who overrides one or both of the methods. The same mutual recursion is added in QueryRunner interface, in this PR.